### ISSUE TITLE
Add more properties to the system partitions in the configuration accessible from the tests

### ISSUE THREAD

**Issue by vkarak** [eth-cscs/reframe#1709]
For the beginning, such properties can be the following: `num_cores`, `num_sockets`, `memory` etc.

This was originally requested by the [EESSI](https://www.eessi-hpc.org/) guys.

**Comment by casparvl** [comment]
@victorusu asked if I could explain a bit more, so that the goal is a bit clearer. Here it goes:

**The idea behind the request**

The primary goal of the EESSI test suite will be to check functionality of the software on a large variety of systems. Secondary goal will be to check e.g. stability of performance over time (per system), so that we can monitor the effect of changes in e.g. the software stack (version updates etc). Getting the test to run completely optimal on a given system is _not_ a goal.

As such, our main requirement is to have tests that require as little tuning as possible to run on different systems (ideally a turn-key test suite). That means defining tests in a way that is as generic as possible. Thus, we prefer system specific things to be as much as possible in `settings.py`. 

**Current situation**

We would like to define a `gromacs.py` ReFrame test that simply spawns one MPI task per core. Currently, the way I would do this is:

```
# gromacs.py
...
self.ntasks = 24
self.ntasks_per_node = 24
```

but this is clearly very specific to my machine. Every site using the EESSI software stack would have to write it's own variant of `gromacs.py`. Even if that will be much easier with the 'library of tests' approach that you are working towards, it's still far from the key-turn test suite we would like to have.

**Requested feature**

It would be very nice if we could specify in the partitions settings.py something like
```
...
'name': 'my_partition',
'scheduler': ...
'cores_per_node': ...
...
```
which could then be used in the tests as 

```
# gromacs.py
...
@rfm.run_after('setup')
def set_num_tasks(self):
    self.ntasks_per_node = self.current_partition.cores_per_node
    self.ntasks = self.ntasks_per_node
```

This way, we could write a generic Gromacs test that runs on any system, on a single node, with one task per core. Again, this might not give the absolute top performance one could potentially achieve on such a system (it might be more performant to launch e.g. one task per socket, or launch 1 task less than the cores_per_node, or... <insert any tuning>), but that's not the primary goal of our test suite. 

**Additional considerations**

We can discuss which partition properties could be useful. In my personal view, it should describe the hardware layout. It should probably define number of `threads`, `cores` and `sockets`. Then, I see two optional sets to define those:

- `sockets_per_node`, `cores_per_socket`, `threads_per_core`
- `sockets_per_node`, `cores_per_node`, `threads_per_node`

Both would provide the possibility of launching e.g. one task per hardware thread, or one task per core - the only difference is if some multiplication/dividision needs to be done first.

I think the most common usage will be e.g.

- `self.ntasks_per_node = self.current_partition.cores_per_node` (pure MPI parallelism)
- `self.ntasks_per_node = self.current_partition.threads_per_node` (pure MPI parallelism, hyperthreading))
- `self.ntasks_per_node = self.current_partition.sockets_per_node` & `self.num_cpus_per_task = self.current_partition.cores_per_socket` (hybrid MPI+openMP, on task per socket)

So there's use cases for both... (an option could be to let the user define `sockets_per_node`, `cores_per_socket`, `threads_per_core` in `settings.py`, but also define `cores_per_node`, `threads_per_node` in the backend for convenience)

**Comment by vkarak** [comment]
Hi @casparvl, it certainly makes sense this feature request. One thing I don't understand very much is what do you mean by this:

> So there's use cases for both... (an option could be to let the user define sockets_per_node, cores_per_socket, threads_per_core in settings.py, but also define cores_per_node, threads_per_node in the backend for convenience)

**Comment by casparvl** [comment]
@vkarak : Sorry, I see how my explaination could be confusing. I was just thinking about how you would like things to be defined in your reframe config file. Either

- `sockets_per_node`, `cores_per_socket`, `threads_per_core`

or

- `sockets_per_node`, `cores_per_node`, `threads_per_node`

Both fully specify the hardware layout, so it's a matter of preference. Of course, ReFrame could support both options in theory (i.e. I could specify `threads_per_core` in my config file, but I could also specify `threads_per_node`), but that might just be a case for confusion: if people erroneously specify _both_, they could conflict. Thus, I was thinking it may be best to enforce _one_ set of valid keywords for the ReFrame config (e.g. `sockets_per_node`, `cores_per_socket`, `threads_per_core`).

Another, seperate configuration, is how one could _use_ these properties in a test. Even with a config that defines `cores_per_socket`, you could still allow people to use _both_ e.g. `self.current_partition.cores_per_node` and `self.current_partition.cores_per_socket` in their ReFrame tests (for convenience). It would mean you'd have to compute `cores_per_node` based on `cores_per_socket` and `sockets_per_node`  in the constructor of `current_partition`. That's a minor detail though, since the user can always compute the multiplication him/herself.

**Comment by vkarak** [comment]
Thanks @casparvl, now I get it. My preference is to allow users to specify this information in a single, unambiguous way. We can then add "convenience" read-only properties to the internal ReFrame object that handles the `current_partition` that will do the math and the mapping.

**Comment by akesandgren** [comment]
sockets/core, cores/socket, threads/core is the more logical way to describe things.

Can I ask for a prio on this? I basically need to have multiple partition defs for our KNL's, with varying threads/core to be able to run tests in the best possible way...
And it would in general be VERY nice to have when designing tests.

**Comment by vkarak** [comment]
> Can I ask for a prio on this?

It's in the pipeline for the next release, so we will soon start actively working on this.

**Comment by akesandgren** [comment]
Excellent!

**Comment by casparvl** [comment]
I have an additional request that was also discussed with @victorusu during the latest EESSI meeting (yesterday).

In some tests, it is useful to know where scratch data can be stored that is local to a node, or that is shared between nodes. In EESSI, this would be used to store a tiered CVMFS cache when trying to run a test with the EESSI container (this is needed to avoid clashes when starting multiple instances of the container on a single node - otherwise their CVMFS local cache dirs clash).

On my system, that might be `/scratch-shared/` and `/scratch-local/`, on your system it might be `/scratch/shared/` and `/scratch/local`. Again, for generalizibility of tests it would be nice if these could be set in the reframe config (I imagine using keywords like `scratch_local` and `scratch_shared` or something), and then used in the tests (e.g. using `self.current_partition.scratch_local`).

I'm not 100% sure if it makes most sense to set this at the system or the partition level: I'm guessing the scratch dirs don't typically vary between partitions in the same system - but if they do, it's nice to be able to set it at the partition level. The disadvantage is a lot of duplication in the `settings.py` if you have a system with a lot of partitions that all have the same scratch dirs...

**Comment by vkarak** [comment]
If I understand this correctly you need to be able to specify the stage directory of tests per system or partition, right? What about the output directory, where test artifacts are now stored? I guess you need to specify this as well.

**Comment by casparvl** [comment]
The ReFrame stage directory you mean? 

I didn't actually consider it, but I guess we could use the staging dir as a shared_scratch, since that has to be a directory shared between nodes anyway. The only downside is that for my specific usage (as a CVMFS cache) it may not be ideal: the stage dir gets cleaned after a successful test, meaning the cache is whiped after every test: not very efficient and it would potentially impact performance a bit, but I guess at least it would work. I can use it in a test through self.current_system.stagedir right?

Also, it also doesn't give me a `local_scratch`, though I guess in 99% of cases, `$TMPDIR` would point to a directory that's suitable as local scratch...

@victorusu what do you think, since you also seemed to think we should add additional config parameters for this. Could we just use the test's staging dir and `$TMPDIR` (so that we don't need to define a seperate `scratch_shared`/`scratch_local` for each partition in the ReFrame config), or am I overlooking something? What would be the added value of a seperate `scratch_shared` over using the ReFrame staging dir?
